rfcs/0001-buildpack-order.md

```diff
@@ -0,0 +1,60 @@
+# Proposed Buildpack Ordering in a Builder
+
+## Summary
+
+`pack inspect-builder gcr.io/paketo-buildpack/builder` should display a `Detection Order` that is deterministic and will never make it impossible for any buildpack to detect.
+
+## Motivation
+
+After a change to the builder detection order it became impossible for `php` apps to detect, because `nginx` and `httpd` were before `php` in the detect order. 
+The `httpd.conf` or `nginx.conf` included in `php` apps will cause the `nginx` or `http` buildpacks to detect `true` before `php` ever gets to run `detect`,
+making it impossible to access the `php` buildpack.
+
+
+## Detailed Explanation
+
+We propose the following rule to order buildpacks in the builder's `Detect Order`:
+
+- All meta buildpacks should be come before implementation buildpacks, and sort each group by buildpaco-id.
+
+- To prove this will always be valid: (1)
+    - Consider 2 buildpacks A and B.
+    - If any of buildpack B's order groups are a subset of any of buildpack A's order groups, then buildpack A must come before buildpack B in the builder order. 
+    - An implementation buildpack can be defined as an order group of itself.
+
+Thus, our exiting buildpacks would be ordered as follows:
+
+- paketo-buildpacks/dotnet-core
+- paketo-buildpacks/go
+- paketo-buildpacks/java
```

Comment:
I think Java triggers on any class files being present. Might want to push this further down? @nebhale

Reply:
Not just _any class files_ the way it was in CF.  It's much more selective these days (hallelujah for modularity and the build plan) and is now existence of any of the following:

* `<APPLICATION_ROOT>/build.gradle`
* `<APPLICATION_ROOT>/pom.xml`
* `<APPLICATION_ROOT>/build.sbt`
* `<APPLICATION_ROOT>/META-INF/MANIFEST.MF`
* `<APPLICATION_ROOT>/WEB-INF/`
* `<APPLICATION_ROOT>/<APPLICATION_NAME>/bin/<APPLICATION_NAME>`

So it can probably go nearly anywhere right now, but I'm not sure that'd be true forever.  Lexically ordering these doesn't solve the fundamental issue of overlapping detection criteria and precedence.